cardano-api/test/Test/Cardano/Api/TxBody.hs

```diff
@@ -0,0 +1,285 @@
+    , txCertificates  = normalizeCertificates  $ txCertificates  content
+    , txIns           = sortOn fst             $ txIns           content
+    , txInsCollateral = normalizeInsCollateral $ txInsCollateral content
+    , txMetadata      = normalizeMetadata      $ txMetadata      content
+    , txMintValue     = normalizeMintValue     $ txMintValue     content
+    , txWithdrawals   = normalizeWithdrawals   $ txWithdrawals   content
+    }
+
+-- | Normalizations applied to roundtrip result data only
+normalizeRoundtrip :: TxBodyContent ViewTx era -> TxBodyContent ViewTx era
+normalizeRoundtrip content@TxBodyContent{txAuxScripts, txIns, txInsCollateral} =
+  content
+    { txAuxScripts    = normalizeAuxScripts    txAuxScripts
+    , txIns           = sortOn fst             txIns
+    , txInsCollateral = normalizeInsCollateral txInsCollateral
+    }
+
+-- | Original data: Unify empty and None.
+--
+-- Roundtrip data: Sort transaction input ids.
+normalizeInsCollateral :: TxInsCollateral era -> TxInsCollateral era
+normalizeInsCollateral = \case
+  TxInsCollateralNone -> TxInsCollateralNone
+  -- for original
+  TxInsCollateral _ ins | null ins -> TxInsCollateralNone
+  -- for roundtrip
+  TxInsCollateral support ins -> TxInsCollateral support $ sort ins
+
+-- | Unify empty and None.
+normalizeMetadata :: TxMetadataInEra era -> TxMetadataInEra era
+normalizeMetadata = \case
+  TxMetadataInEra _ (TxMetadata m) | null m -> TxMetadataNone
+  other                                     -> other
+
+-- | Unify empty and None.
+-- Upgrade script versions (see Upgrading scripts section).
+-- Stabilize order of scripts sorting them by language.
+normalizeAuxScripts :: TxAuxScripts era -> TxAuxScripts era
+normalizeAuxScripts = \case
+  TxAuxScripts _       []      -> TxAuxScriptsNone
+  TxAuxScripts support scripts ->
+    -- sorting uses script versions, hence sort after upgrade
+    TxAuxScripts support $
+    sortOn languageOfScriptInEra $ map upgradeScriptInEra scripts
+  other -> other
+
+languageOfScriptInEra :: ScriptInEra era -> AnyScriptLanguage
+languageOfScriptInEra (ScriptInEra lang _) =
+  AnyScriptLanguage $ languageOfScriptLanguageInEra lang
+
+-- | Unify empty and None.
+normalizeWithdrawals :: TxWithdrawals ViewTx era -> TxWithdrawals ViewTx era
+normalizeWithdrawals = \case
+  TxWithdrawals _ [] -> TxWithdrawalsNone
+  other              -> other
+
+-- | Unify empty and None.
+normalizeCertificates :: TxCertificates ViewTx era -> TxCertificates ViewTx era
+normalizeCertificates = \case
+  TxCertificates _ [] _ -> TxCertificatesNone
+  other                 -> other
+
+-- | Unify empty and None.
+normalizeMintValue :: TxMintValue ViewTx era -> TxMintValue ViewTx era
+normalizeMintValue = \case
+  TxMintValue _ v _ | v == mempty -> TxMintNone
+  other                           -> other
+
+
+-- * Ugrading scripts
+--
+-- The instruction set from V1 may be used as V2,
+-- and we can't determine the language version from the transaction.
```

Comment:
We should change this comment to: "We can't determine the script language version from the ledger transaction. Therefore we default to the highest version for the time being. TODO: Revisit this"